pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,259 @@
+		if !tokenResult.Status.Authenticated {
+			klog.Warningf("Token for %s/%s did not authenticate correctly", saNamespace, saName)
+			return false, nil
+		}
+		if tokenResult.Status.User.Username != username {
+			klog.Warningf("Token for %s/%s authenticated as unexpected username: %s", saNamespace, saName, tokenResult.Status.User.Username)
+			return false, nil
+		}
+		klog.V(4).Infof("Verified credential for %s/%s", saNamespace, saName)
+		return true, nil
+	}
+
+	// If we couldn't run the token review, the API might be disabled or we might not have permission.
+	// Try to make a request to /apis with the token. If we get a 401 we should consider the token invalid.
+	clientConfigCopy := *clientConfig
+	clientConfigCopy.BearerToken = token
+	clientConfigCopy.NegotiatedSerializer = legacyscheme.Codecs
+	client, err := restclient.UnversionedRESTClientFor(&clientConfigCopy)
+	if err != nil {
+		return false, err
+	}
+	err = client.Get().AbsPath("/apis").Do().Error()
+	if apierrors.IsUnauthorized(err) {
+		klog.Warningf("Token for %s/%s did not authenticate correctly: %v", saNamespace, saName, err)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (t *DynamicControllerClientBuilder) getOrCreateServiceAccount(name string) (*v1.ServiceAccount, error) {
+	sa, err := t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	if err == nil {
+		return sa, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); err != nil {
+		t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}})
+	}
+
+	// Create the service account
+	sa, err = t.CoreClient.ServiceAccounts(t.Namespace).Create(&v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Namespace: t.Namespace, Name: name}})
+	if apierrors.IsAlreadyExists(err) {
+		// If we're racing to init and someone else already created it, re-fetch
+		return t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	}
+	return sa, err
+}
+
+func (t *DynamicControllerClientBuilder) startSync() {
+	period := defaultExpiration * 20 / 100 / 10
+
+	go wait.Until(func() {
+		t.configs.Range(func(key, value interface{}) bool {
+			name := key.(string)
+			slot := value.(*configSlot)
+			now := t.clock.Now()
+
+			// Require a refresh if within 20% of the expiration time.
+			iat := slot.expTime.Time.Add(-1 * time.Duration(defaultExpiration*20/100) * time.Second)
+
+			if now.After(iat) {
+				token, exp, err := t.getToken(name, slot.config)
+				if err != nil {
+					// We endure error here since we have 10 chances to rotate token
+					klog.Warningf("get service account token failed for %s/%s: %v", t.Namespace, name, err)
+				}
+				slot.config.BearerToken = token
```

Comment:
modifying this field this way is not threadsafe, and will not propagate the new token to the transport that consumed the token:

https://github.com/kubernetes/kubernetes/blob/7ca9d1bdd878286a7d273e61e92a16f680cfbce7/staging/src/k8s.io/client-go/transport/round_trippers.go#L47-L52